pyqtorch/circuit.py

```diff
@@ -43,46 +43,71 @@ def __eq__(self, other: Any) -> bool:
-        if self.diff_mode == DiffMode.AD:
-            state = self.run(state, values)
-            return overlap(state, observable.forward(state, values))
+    @property
+    def _device(self) -> device | None:
+        devices = set()
+        for op in self.operations:
+            if isinstance(op, QuantumCircuit):
+                devices.add(op._device)
+            elif isinstance(op, Module):
+                devices.update([b.device for b in op.buffers()])
+        if len(devices) == 1 and None not in devices:
+            _device = next(iter(devices))
+            logger.debug(f"Found device {_device}.")
+            return _device
         else:
-            from pyqtorch.adjoint import AdjointExpectation
-
-            return AdjointExpectation.apply(
-                self, observable, state, values.keys(), *values.values()
+            logger.warning(
+                f"Unable to determine device of module {self}.\
+                             Found {devices}, however expected exactly one device."
             )
+            return None
 
-    @property
-    def _device(self) -> torch.device:
-        try:
-            (_, buffer) = next(self.named_buffers())
-            return buffer.device
-        except StopIteration:
-            return torch.device("cpu")
-
-    def init_state(self, batch_size: int = 1) -> torch.Tensor:
+    def init_state(self, batch_size: int = 1) -> Tensor:
         return zero_state(self.n_qubits, batch_size, device=self._device)
 
     def reverse(self) -> QuantumCircuit:
-        return QuantumCircuit(self.n_qubits, torch.nn.ModuleList(list(reversed(self.operations))))
+        return QuantumCircuit(self.n_qubits, ModuleList(list(reversed(self.operations))))
+
+    def to(self, device: device) -> QuantumCircuit:
+        self.operations = ModuleList([op.to(device) for op in self.operations])
+        return self
+
+
+def expectation(
+    circuit: QuantumCircuit,
+    state: State,
+    values: dict[str, Tensor],
+    observable: QuantumCircuit,
+    diff_mode: DiffMode = DiffMode.AD,
+) -> Tensor:
+    """Compute the expectation value of the circuit given a state and observable.
+    Arguments:
+        circuit: QuantumCircuit instance
+        state: An input state
+        values: A dictionary of parameter values
+        observable: QuantumCircuit representing the observable
+        diff_mode: The differentiation mode
+    Returns:
+        A expectation value.
+    """
+    if observable is None:
+        raise ValueError("Please provide an observable to compute expectation.")
+    if state is None:
+        state = circuit.init_state(batch_size=1)
+    if diff_mode == DiffMode.AD:
+        state = circuit.run(state, values)
+        return overlap(state, observable.forward(state, values))
+    else:
+        from pyqtorch.adjoint import AdjointExpectation
+
+        return AdjointExpectation.apply(circuit, observable, state, values.keys(), *values.values())
```

Comment:
if the user erroneously passes `DiffMode.GPSR` this will not result in the expected outcome. 
I'd say in that case you also raise `ValueError`.